tcti/test: switch to swtpm

joholl opened the pull request:
The currently used simulator `mssim` by IBM has some downsides. The main problem is its lack of open source character:

- hosted on SourceForge
- only inconvenient ticket system, long response times
- no possibility to contribute (besides uploading a patch file in a ticket)
- no version control at all (I could not find a git repository)

Heck, the build process for linux is documented in a Microsoft Word file. Basically, there is no way for the community to contribute to this vital part of the TPM software ecosystem.

As an alternative, I propose using the [stefanberger/swtpm](https://github.com/stefanberger/swtpm) simulator.

**EDIT**:
Like the mssim, swtpm provides two sockets for communication. However, the swtpm's control socket follows a different protocol than the mssim's platform socket. For testing on mssim, two tests depend on the mssim-specific function `tcti_platform_command()`:

* `tpmclient`: resets the TPM a few times
* `sapi_command_cancel`: cancels a command using `tcti_platform_command()`   
(does not call `Tss2_Tcti_Cancel()`).

I added a new TCTI function `Tss2_Tcti_Reset()`. For tbs and device it will return `TSS2_TCTI_RC_NOT_IMPLEMENTED`. I don't know how exactly the TCTI versioning works, for now I added it to `TSS2_TCTI_CONTEXT_COMMON_V2`.

IMHO, the `sapi_command_cancel` test does not really test the TSS but simulator functionality. I think it would only make sense if it called `Tss2_Tcti_Cancel()`. The swtpm, however, does not support cancel, yet.

The first commit is just for switching to my own DockerHub where I added the dependencies for the swtpm. See my changes for the dockerfiles: joholl/tpm2-software-container@9e5781f8026. I did not remove any mssim dependencies. If you agree, I'll submit a PR for tpm2-software-container and rebase this PR.

**EDIT2**:
A good amount of work was done by @AndreasFuchsSIT who started working on this on his fork already.

joholl commented:
I'm not sure what the problem with LGTM is... The error is `could not build the merge commit` and it failed due to the missing swtpm headers (which I have installed in my DockerHub).

diabonas commented:
I like the idea of using a TPM simulator that is developed more actively and transparently. Have you considered using [ms-tpm-20-ref](https://github.com/microsoft/ms-tpm-20-ref) as an alternative? It is Microsoft's own packaging of the code that [ibmswtpm](https://sourceforge.net/projects/ibmswtpm2/) is based on, hence its interface is compatible with `tpm_server`. The project is hosted on GitHub licensed under a BSD license.

I have managed to run the complete test suite by changing the call of the simulator executable from `tpm_server -port <port>` to `tpm_server <port>`, so switching to ms-tpm-20-ref would be very easy. I can see only two downsides:
- The project currently has no tagged releases.
- The Linux build has been broken multiple times (https://github.com/microsoft/ms-tpm-20-ref/commit/085d79278681838b94376ebc15ca58a58824d684, https://github.com/microsoft/ms-tpm-20-ref/commit/eeaafeab0a3503de3125d8f7a4362567cfbf823d), but currently works. There are some comments on the latest commit https://github.com/microsoft/ms-tpm-20-ref/commit/39e7306e235a6b2e09e454af08de285388e7a4b4 that should help in permanently fixing these build problems in the future.

---

Regarding the LGTM build failures, lgtm.com has its own build environment that unfortunately cannot use our Dockerfile. You have to build and install libtpms and swtpm from source by modifying [`.lgtm.yml`](https://github.com/tpm2-software/tpm2-tss/blob/master/.lgtm.yml), see e.g. the [tpm2-tools LGTM configuration](https://github.com/tpm2-software/tpm2-tools/blob/master/.lgtm.yml) for an example how this can be done.

diabonas commented:
I have opened draft PR https://github.com/tpm2-software/tpm2-tss/pull/1544 to show how to use ms-tpm-20-ref as a drop-in replacement for the current simulator.

joholl commented:
Good call! Actually I did indeed play around with the ms-tpm-20-ref simulator earlier this year (see joholl/tpm2-tss@872716727c47fa68d4579ac634b72ba3e9a95fdc). I agree that it's much easier to replace the IBM simulator by ms-tpm-20-ref.

The problem with the ms-tpm-20-ref is the [CLA](https://opensource.microsoft.com/pdf/microsoft-contribution-license-agreement.pdf) which has to be sent to MS beforehand. IMO this is in the best case annoying and in the worst a real obstacle (especially for developers working for companies with legal departments...). I think this CLA is a stumbling block for any potential contributor.

diabonas commented:
> The problem with the ms-tpm-20-ref is the [CLA](https://opensource.microsoft.com/pdf/microsoft-contribution-license-agreement.pdf) which has to be sent to MS beforehand. IMO this is in the best case annoying and in the worst a real obstacle (especially for developers working for companies with legal departments...). I think this CLA is a stumbling block for any potential contributor.

Good point, I agree that CLAs can be annoying, especially if you are representing a company. I didn't see that because my PR https://github.com/microsoft/ms-tpm-20-ref/pull/39 didn't require me to sign the CLA, apparently because it is only necessary for contributions exceeding a certain size, see https://github.com/microsoft/ms-tpm-20-ref/issues/9#issuecomment-397191977. @flihp on the other hand apparently really had to do so for https://github.com/microsoft/ms-tpm-20-ref/pull/8.

AndreasFuchsSIT commented:
I have to say, I like the idea and swtpm seems to be much better in shape than the other two...
@joholl please go ahead and develop this further. I'd propose patch1: tcti-swtpm and patch2: switch CI to swtpm.

The LGTM-Error is on `src/tss2-tcti/tcti-swtpm.h:13:10: fatal error: swtpm/tpm_ioctl.h: No such file or directory`
What is swtpm/tpm_ioctl.h actually used for ? esp since it's part of the header ?

@diabonas I appreciate the ms-tpm PR which is already an improvement. I hope you're ok with going for swtpm which IMHO is even better ?

diabonas commented:
@AndreasFuchsSIT I don't mind at all, swtpm is a fine choice! I just wanted to quickly show the possible alternative, since it was practically zero effort anyway :)

I can have a look into modifying the LGTM configuration later this week, shouldn't be too difficult to get this to work I hope.

AndreasFuchsSIT commented:
@diabonas Thanks for looking into this
@joholl do we need the swtpm/tpm_ioctl.h anyways and what for ?

diabonas commented on configure.ac:
```diff
@@ -138,6 +138,14 @@ AM_CONDITIONAL([ENABLE_TCTI_MSSIM], [test "x$enable_tcti_mssim" != xno])
+            [enable_tcti_swtpm=yes])
+AM_CONDITIONAL([ENABLE_TCTI_SWTPM], [test "x$enable_tcti_swtpm" != xno])
+AS_IF([test "x$enable_tcti_swtpm" = "xyes"],
+	AC_DEFINE([TCTI_SWTPM],[1], [TCTI FOR SWTPM]))
```
We need to check if swtpm is installed, or at least whether the required header file is present, otherwise configure can succeed, but compilation fails with
```
src/tss2-tcti/tcti-swtpm.h:13:10: fatal error: swtpm/tpm_ioctl.h: No such file or directory
```
Unfortunately swtpm doesn't include a pkg-config file, so we'll need
```
AC_CHECK_HEADER([swtpm/tpm_ioctl.h], , AC_MSG_ERROR([Missing header swtpm/tpm_ioctl.h: install swtpm.]))
```
probably directly here in the `AS_IF`?

joholl replied:
Yes, if we use the header, we should add a check. I'll with my next push. Thanks!

diabonas commented:
@AndreasFuchsSIT [`swtpm/tpm_ioctl.h`](https://github.com/stefanberger/swtpm/blob/master/include/swtpm/tpm_ioctl.h) appears to be necessary to define some data types used in `tss2-tcti/tcti-swtpm.c` (the ones prefixed with `ptm_`/`PTM_`).

@joholl Since only the header file is necessary for the build (LGTM doesn't actually run the test suite, in fact our current configuration doesn't even attempt to build it), the following `.lgtm.yml` should work:
```
extraction:
  cpp:
    prepare:
      packages: "autoconf-archive"
    after_prepare:
    - cd "$LGTM_WORKSPACE"
    - wget https://github.com/stefanberger/swtpm/archive/v0.2.0.tar.gz
    - tar xf v0.2.0.tar.gz
    - export C_INCLUDE_PATH="$LGTM_WORKSPACE/swtpm-0.2.0/include:$C_INCLUDE_PATH"
```

joholl commented:
First up: thanks for the feedback @AndreasFuchsSIT @diabonas!

In my next push, I'll gladly split into two commits a) introduce tcti-swtpm and b) switch CI to swtpm. I'll also update the LGTM file and remove the change of permissions which I didn't notice.

The header `swtpm/tpm_ioctl.h` is actually only used for the control command codes  `CMD_INIT` and `CMD_SET_LOCALITY`. Assuming they will not change anytime soon, we could copy these defines. In this case we could even use Fedora's packaged version of swtpm. As for the `#include`: I'll move it from `tcti-swtpm.c` to `tcti-swtpm.h`.

joholl commented on lib/tss2-tcti-swtpm.pc.in:
```diff
@@ -0,0 +1,12 @@
+Description: TCTI library for communicating with swtpm.
+URL: https://github.com/tpm2-software/tpm2-tss
+Version: @VERSION@
+Requires: tss2-mu
```
Thanks, I was unaware of that.

joholl commented on lib/tss2-tcti-swtpm.pc.in:
```diff
@@ -0,0 +1,12 @@
+Version: @VERSION@
+Requires: tss2-mu
+Cflags: -I${includedir}
+Libs: -ltss2-tcti-mssim -L${libdir}
```
You're right, copy-pasted and missed.

joholl commented:
Changes:
- Removed dependency on swtpm header `swtpm/tpm_ioctl.h`. Thus, I did *not* add the dependency in `configure.ac` and `.lgtm.yml`.
- Split into three commits `add tcti-swtpm`, `tcti: add funtion Tss2_Tcti_Reset() to TCTI V2`, and ` ci: switch to swtpm`
- incorporated feedback about `lib/tss2-tcti-swtpm.pc.in`
- add various tests to increase coverage
- minor misc changes

LGTM succeeds now. I did not change `.lgtm.yml` but removed the dependency on the swtpm headers.

To be honest, I'm at a loss why codecov rejects my PR.

```
codecov/patch — 68.93% of diff hit (target 84.08%) 
```

The biggest portion of added code is located in `src/tss2-tcti/tcti-swtpm.c` with an apparent coverage of **68.85 %**. However, when I run `make check-code-coverage` locally, I get **86.7 % (208 / 240)** for the coverage of `src/tss2-tcti/tcti-swtpm.c`.

Today, I updated the tests (mainly in da3740f) and somehow codecov seems to fail using the updated test-suite. When I have a look at the [coverage](https://codecov.io/gh/tpm2-software/tpm2-tss/compare/30afabfc81496bc8a225e67b4d090ba1e169a22a...1158f71d193c643baa24ca8ec453b3c8102c7111/diff) of `src/tss2-tcti/tcti-swtpm.c`, I notice uncovered lines which clearly should be covered.

The most obvious case is the uncovered function `Tss2_Tcti_Info()` in `src/tss2-tcti/tcti-swtpm.c:614`. This should clearly be covered by [this test](https://github.com/tpm2-software/tpm2-tss/blob/da3740fa15e7b774dfcded21249f3c293724d0c7/test/unit/tcti-swtpm.c#L268) added in da3740fa15e7b774dfcded21249f3c293724d0c7.

joholl commented:
## NO NOT MERGE YET

The first commit is just for switching to my own DockerHub where I added the dependencies for the swtpm. See my changes for the dockerfiles: joholl/tpm2-software-container@9e5781f. I did not remove any mssim dependencies. If you agree, I'll submit a PR for tpm2-software-container and rebase this PR.

AndreasFuchsSIT commented on src/tss2-tcti/tcti-swtpm.c:
```diff
@@ -0,0 +1,640 @@
+        return rc;
+    }
+
+    if (timeout != TSS2_TCTI_TIMEOUT_BLOCK) {
```
Could we get async I/O working ?
That's gonna be needed by FAPI and potentially others in the future.

joholl replied:
AFAIK, `libtpms` does not support async rx. It does not even support separated tx/rx but only one call 
```c
TPM_RESULT TPMLIB_Process(unsigned char **respbuffer, uint32_t *resp_size,
                          uint32_t *respbufsize,
                          unsigned char *command, uint32_t command_size);
```

AndreasFuchsSIT commented on include/tss2/tss2_tcti.h:
```diff
@@ -83,6 +83,8 @@ typedef void TSS2_TCTI_POLL_HANDLE;
     ((TSS2_TCTI_CONTEXT_COMMON_V1*)tctiContext)->setLocality
 #define TSS2_TCTI_MAKE_STICKY(tctiContext) \
     ((TSS2_TCTI_CONTEXT_COMMON_V2*)tctiContext)->makeSticky
+#define TSS2_TCTI_RESET(tctiContext) \
```
Sorry, but this is TCG-spec.
We cannot just add things in here, esp if they are very simulator specific rather than general-purpose-TPM.

We can add a `Tss2_Tcti_swtpm_Reset()` to tss2_tcti_swtpm.h for this.

joholl replied:
Oh ok. That's what I suspected. Do you mind if the tests are not TCTI-agnostic? If so, I'll replace mssim-specific commands by swtpm-specific commands.

AndreasFuchsSIT replied:
If it's unit test, I don't care, cause they are tcti-specific anyways.
If it's integration tests, I do care, but I cannot remember an int-test to use this anyways (since it was not exported by the .map file)...

joholl replied:
I'm talking about integration tests. Previously, `tpmclient` and `sapi_command_cancel` call the mssim-specific function `tcti_platform_command()`.
`tpmclient` resets the TPM and `sapi_command_cancel` cancels commands (without using the dedicated TCTI function) and is IMO anyway obsolete.

joholl commented on src/tss2-tcti/tcti-swtpm.c:
```diff
@@ -0,0 +1,640 @@
+
+/*
+ * For compatibility to the mssim platform commands issued by some tests:
+ * Mapping mssim platform commands to swtpm control channel commands //TODO
```
This is code is actually outdated. After squashing commits 2 and 3, it's gone :)

joholl commented on src/tss2-tcti/tcti-swtpm.c:
```diff
@@ -0,0 +1,640 @@
+ * For compatibility to the mssim platform commands issued by some tests:
+ * Mapping mssim platform commands to swtpm control channel commands //TODO
+ */
+TSS2_RC tcti_platform_command (
```
Again, outdated. I added a new TCTI function Tss2_Tcti_Reset() in the commit you asked me to squash. I'm just not sure if I did it the right way (which is why I did not squash, yet)

joholl commented on src/tss2-tcti/tcti-swtpm.c:
```diff
@@ -0,0 +1,640 @@
+    if (tcti_swtpm == NULL) {
+        return TSS2_TCTI_RC_BAD_CONTEXT;
+    }
+
```
If `cmd_sdu` is `NULL`, `cmd_sdu_len` will never be used. Do you think I should add it anyway for good style?

AndreasFuchsSIT replied:
For input consistency checking... ;-)

joholl commented on src/tss2-tcti/tcti-swtpm.c:
```diff
@@ -0,0 +1,640 @@
+        return TSS2_TCTI_RC_BAD_CONTEXT;
+    }
+
+    if (resp_sdu != NULL && resp_sdu_len == NULL) {
```
Good call

diabonas commented on configure.ac:
```diff
@@ -182,7 +182,7 @@ AC_ARG_ENABLE([integration],
     [enable_integration=no])
 AS_IF([test "x$enable_integration" = "xyes"],
-      [ERROR_IF_NO_PROG([tpm_server])
+      [ERROR_IF_NO_PROG([swtpm])
```
The test in [l. 199](https://github.com/tpm2-software/tpm2-tss/blob/30afabfc81496bc8a225e67b4d090ba1e169a22a/configure.ac#L199) for the mssim TCTI module needs to be changed to swtpm as well.

joholl replied:
Thanks, solved this by the latest changes.

diabonas commented on Makefile-test.am:
```diff
@@ -16,9 +16,11 @@ if TESTPTPM
 INT_LOG_FLAGS = --ptpm=$(PTPM)
 else
-INT_LOG_COMPILER = $(srcdir)/script/int-log-compiler.sh
+INT_LOG_COMPILER = $(srcdir)/script/int-log-compiler-swtpm.sh
```
As long as we still support tpm_server, it might be nice to use `AC_CHECK_PROG` whether tpm_server or swtpm are available and switch the log compiler accordingly. Depends a bit on how long we want to keep tpm_server around whether it's worth the effort...

joholl replied:
Good call. I added  a mechanism which chooses the testing backend. I also streamlined the code for choosing `device` (formerly `ptpm`) with this new mechanism.

In configure.ac, I added two output variables to configure.ac:

  - INTEGRATION_TCTI: `mssim`, `swtpm` or `device`   
(according to this variable, int-log-compiler-<...>.sh will be called for testing)
  - INTEGRATION_ARGS: arguments to passed to int-log-compiler script   
(`--device=<...>` for device, empty otherwise)

 The mechanism for choosing the testing backend is as follows:

  - if --with-device=<..> was given:
     - INTEGRATION_TCTI = `device`
     - INTEGRATION_ARGS = `--device=<..>`

  - if tcti-swtpm is enabled and swtpm in PATH
     - INTEGRATION_TCTI = `swtpm`

  - if tcti-mssim is enabled and tpm_server in PATH
     - INTEGRATION_TCTI = `mssim`

   - if no backend could be found: error

joholl commented:
Thank you for your feedback. I changed my code as for your requests. There are 2 major things I changed:

- Add a mechanism which automatically chooses the testing backend (int-log-compiler). I also streamlined the code for choosing `device` (formerly `ptpm`) with this new mechanism.
- Remove test `sapi-cancel-command` since it does not test TSS code.

There is still a **bug** I found which is related to this PR. If `--disable-tcti-mssim` is passed, `tpmclient` does not build since it depends on `mssim`. This also applies to swtpm now, since it depends now on both swtpm and mssim for resetting the TPM. Any ideas?

joholl commented on src/tss2-tcti/tcti-swtpm.c:
```diff
@@ -0,0 +1,618 @@
+        LOG_DEBUG ("response size: %" PRIu32, tcti_common->header.size);
+    }
+
+    *response_size = tcti_common->header.size;
```
Good find! Thank you for pointing that out. Fixed now.

AndreasFuchsSIT commented on .ci/travis.run:
```diff
@@ -22,7 +22,7 @@ if [ "$TRAVIS_BRANCH" != "coverity_scan" ]; then
   ci_env=$(bash <(curl -s https://codecov.io/env))
   docker run $ci_env --env-file .ci/docker.env \
-    -v "$(pwd):/workspace/tpm2-tss" "tpm2software/tpm2-tss:$DOCKER_TAG" \
+    -v "$(pwd):/workspace/tpm2-tss" "joholl/tpm2-tss:$DOCKER_TAG" \
```
Reset to tpm2software

AndreasFuchsSIT commented on .ci/travis.run:
```diff
@@ -52,7 +52,7 @@ mv cov-analysis-* cov-analysis
 # perform the scan
 docker run --env-file .ci/docker.env \
-  -v "$(pwd):/workspace/tpm2-tss" "tpm2software/tpm2-tss:$DOCKER_TAG" \
+  -v "$(pwd):/workspace/tpm2-tss" "joholl/tpm2-tss:$DOCKER_TAG" \
```
Same

AndreasFuchsSIT commented on Dockerfile:
```diff
@@ -1,4 +1,4 @@
-FROM tpm2software/tpm2-tss AS base
+FROM joholl/tpm2-tss AS base
```
Same

joholl commented:
Done. I rebased and applied the changes you made to mssim also to swtpm. Travis succeeds.

tstruk commented on src/tss2-tcti/tcti-swtpm.c:
```diff
@@ -0,0 +1,632 @@
+    }
+
+    socket_close (&tcti_swtpm->tpm_sock);
+}
```
finalize function will also need to free the tcti_swtpm->conf_copy allocated in Init.

joholl replied:
Good call! Thanks :)

tstruk commented:
> @tstruk Merge ?
I added one comment. This will also need to be rebased. @joholl could you do that please. Let's merge it after 2.4.0

joholl commented:
Fixed and rebased

pdxjohnny commented:
Will this be in 3.0?

tstruk commented:
> Will this be in 3.0?
Yes

tstruk commented:
Newer version of the PR is here: https://github.com/tpm2-software/tpm2-tss/pull/1654
